application/src/test/java/org/togetherjava/tjbot/jda/ButtonClickEventBuilder.java

```diff
@@ -0,0 +1,245 @@
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import net.dv8tion.jda.api.MessageBuilder;
+import net.dv8tion.jda.api.entities.Member;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.entities.MessageEmbed;
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.events.interaction.ButtonClickEvent;
+import net.dv8tion.jda.api.interactions.components.ActionRow;
+import net.dv8tion.jda.api.interactions.components.Button;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.togetherjava.tjbot.commands.SlashCommand;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+import java.util.stream.Stream;
+
+import static org.mockito.Mockito.when;
+
+/**
+ * Builder to create button click events that can be used for example with
+ * {@link SlashCommand#onButtonClick(ButtonClickEvent, List)}.
+ * <p>
+ * Create instances of this class by using {@link JdaTester#createButtonClickEvent()}.
+ * <p>
+ * Among other Discord related things, the builder optionally accepts a message
+ * ({@link #message(Message)}) and the user who clicked on the button
+ * ({@link #userWhoClicked(Member)} ). As well as several ways to modify the message directly for
+ * convenience, such as {@link #content(String)} or {@link #actionRows(ActionRow...)}. The builder
+ * is by default already setup with a valid dummy message and the user who clicked the button is set
+ * to the author of the message.
+ * <p>
+ * In order to build the event, at least one button has to be added to the message and marked as
+ * <i>clicked</i>. Therefore, use {@link #actionRows(ActionRow...)} or modify the message manually
+ * using {@link #message(Message)}. Then mark the desired button as clicked using
+ * {@link #build(Button)} or, if the message only contains a single button,
+ * {@link #buildWithSingleButton()} will automatically select the button.
+ * <p>
+ * Refer to the following examples:
+ *
+ * <pre>
+ * {@code
+ * // Default message with a delete button
+ * jdaTester.createButtonClickEvent()
+ *   .actionRows(ActionRow.of(Button.of(ButtonStyle.DANGER, "1", "Delete"))
+ *   .buildWithSingleButton();
+ *
+ * // More complex message with a user who clicked the button that is not the message author and multiple buttons
+ * Button clickedButton = Button.of(ButtonStyle.PRIMARY, "1", "Next");
+ * jdaTester.createButtonClickEvent()
+ *   .message(new MessageBuilder()
+ *     .setContent("See the following entry")
+ *     .setEmbeds(
+ *       new EmbedBuilder()
+ *         .setDescription("John")
+ *         .build())
+ *     .build())
+ *   .userWhoClicked(jdaTester.createMemberSpy(5))
+ *   .actionRows(
+ *     ActionRow.of(Button.of(ButtonStyle.PRIMARY, "1", "Previous"),
+ *     clickedButton)
+ *   .build(clickedButton);
+ * }
+ * </pre>
+ */
+public final class ButtonClickEventBuilder {
```

Comment:
It's not a record, these methods are setters? I don't understand where the naming is from

They should be named "set" so it's clear what they do, currently I'd think they all *add* to the existing value.
But they set, not add, they replace the old value.

Reply:
Kinda depends on which builder pattern you are following. There are builder patterns that use `setXXX` and some that just use `xxx` (Intellijs built-in builder-generator for example). Records also use just `xxx`.

Reply:
Records aren't builders, that's a huge difference.

And I just feel like we've to be clear about it's behaviour, either add it in docs or make it clear in the method's name

Reply:
As said, this is not just records. It is actually a common pattern. Lombdok and Intellijs default settings create builders like that. Maybe you are not aware of this convention or dont like it, thats fair - but it is common.

At the same time, `setXXX` and `withXXX` are also common 🤷 

Personally, I do not care that much and am willing to change it if it helps you. Just want to make sure that its clear that all 3 are commonly used patterns for builders.

Reply:
Renamed it to `setXXX` now.